### PR TITLE
drivers: opt3001: adding threshold interrupt.

### PR DESCRIPTION
**Pull request by GiulianoFranchetto**
This commit adds interrupt management on the opt300x sensors. The
only available interrupt is a lower/upper threshold, as defined in
the component datasheet.

Signed-off-by: GiulianoFranchetto <giuliano.franchetto@intellinium.com>

### CODE REVIEW

File: drivers/sensor/opt3001/Kconfig [hunk 1]
```diff
@@ -3,8 +3,53 @@
...
 # SPDX-License-Identifier: Apache-2.0
 
-config OPT3001
+menuconfig OPT3001
 	bool "OPT3001 Light Sensor"
 	depends on I2C
 	help
 	  Enable driver for OPT3001 light sensors.
+
+if (OPT3001)
```
Comment (galak) [inline drivers/sensor/opt3001/Kconfig:12]: nit remove parens, not need, and we don't generally do this.

File: drivers/sensor/opt3001/Kconfig [hunk 2]
```diff
@@ -3,8 +3,53 @@
...
+	bool "Use own thread"
+	depends on GPIO
+	select OPT3001_TRIGGER
+
+endchoice
+
+config OPT3001_TRIGGER
+	bool
+
+if (OPT3001_TRIGGER)
```
Comment (galak) [inline drivers/sensor/opt3001/Kconfig:37]: remove parens.

File: drivers/sensor/opt3001/opt3001.c [hunk 3]
```diff
@@ -22,11 +22,11 @@ static int opt3001_reg_read(struct opt3001_data *drv_data, uint8_t reg,
 	uint8_t value[2];
 
 	if (i2c_burst_read(drv_data->i2c, DT_INST_REG_ADDR(0),
-		reg, value, 2) != 0) {
+			   reg, value, 2) != 0) {
 		return -EIO;
 	}
 
-	*val = ((uint16_t)value[0] << 8) + value[1];
+	*val = ((uint16_t) value[0] << 8) + value[1];
```
Comment (MaureenHelm) [inline drivers/sensor/opt3001/opt3001.c:29]: Please move unrelated whitespace/formatting changes to a separate commit

File: drivers/sensor/opt3001/opt3001.c [hunk 4]
```diff
@@ -39,7 +39,7 @@ static int opt3001_reg_write(struct opt3001_data *drv_data, uint8_t reg,
 	new_value[0] = val >> 8;
 	new_value[1] = val & 0xff;
 
-	uint8_t tx_buf[3] = { reg, new_value[0], new_value[1] };
+	uint8_t tx_buf[3] = {reg, new_value[0], new_value[1]};
```
Comment (MaureenHelm) [inline drivers/sensor/opt3001/opt3001.c:42]: More unrelated whitespace changes

File: drivers/sensor/opt3001/opt3001_trigger.c [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+	uint16_t config;
+	struct opt3001_data *data = dev->data;
+	const struct opt3001_config *cfg = dev->config;
+	struct sensor_trigger limit_trig = {
+		.type = SENSOR_TRIG_THRESHOLD,
+		.chan = SENSOR_CHAN_ALL,
+	};
+
+	/* Reading the configuration field to clear irq */
+	err = i2c_burst_read(data->i2c, DT_INST_REG_ADDR(0), 0x01,
```
Comment (MaureenHelm) [inline drivers/sensor/opt3001/opt3001_trigger.c:45]: Use `i2c_burst_read_dt` instead

File: drivers/sensor/opt3001/opt3001_trigger.c [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+int opt3001_set_th(const struct device *dev,
+		   uint16_t value,
+		   uint8_t addr)
+{
+	uint8_t le;
+	uint16_t tl;
+	uint32_t v;
+	struct opt3001_data *data = dev->data;
+
+	if (value <= 40.95) {
```
Comment (MaureenHelm) [inline drivers/sensor/opt3001/opt3001_trigger.c:91]: Please use some self-documenting preprocessor macros for these magic numbers

File: drivers/sensor/opt3001/opt3001_trigger.c [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+	tl = (v & 0x0FFF) | (le << 12);
+	tl = sys_be16_to_cpu(tl);
+
+	uint8_t buf[3] = {
+		addr,
+		tl & 0xff,
+		tl >> 8
+	};
+
+	return i2c_write(data->i2c, buf, 3, DT_INST_REG_ADDR(0));
```
Comment (MaureenHelm) [inline drivers/sensor/opt3001/opt3001_trigger.c:127]: Use `i2c_write_dt` instead

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/opt3001/opt3001.c [hunk 8]
```diff
@@ -61,6 +62,50 @@ static int opt3001_reg_update(const struct device *dev, uint8_t reg,
...
+	return opt3001_set_th(dev, value->val1, 0x02);
+}
+
+static int opt3001_attr_set_high_th(const struct device *dev,
+				    const struct sensor_value *value)
+{
+	return opt3001_set_th(dev, value->val1, 0x03);
+}
+
+static int opt3001_attr_set_light(const struct device *dev,
```
Comment (cfriedt) [inline drivers/sensor/opt3001/opt3001.c:78]: Can this be renamed to `opt3001_attr_set_threshold`?

Comment (GiulianoFranchetto) [reply drivers/sensor/opt3001/opt3001.c]: Renaming done.

**Review by cfriedt — APPROVED** [review]

File: drivers/sensor/opt3001/opt3001_trigger.c [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+		le = 5;
+	} else if (1310.40 < value && value <= 2620.80) {
+		le = 6;
+	} else if (2620.80 < value && value <= 5241.60) {
+		le = 7;
+	} else if (5241.60 < value && value <= 10483.20) {
+		le = 8;
+	} else if (10483.20 < value && value <= 20966.40) {
+		le = 9;
+	} else if (20966.40 < value && value <= 41932.80) {
```
Comment (MaureenHelm) [inline drivers/sensor/opt3001/opt3001_trigger.c:116]: The first part of these expressions are redundant (you don't need the value is greater than checks)

File: drivers/sensor/opt3001/opt3001_trigger.c [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+
+	gpio_pin_interrupt_configure_dt(&config->irq_spec,
+					GPIO_INT_DISABLE);
+
+	err = opt3001_set_lower_th(dev, 0x0000);
+	if (err) {
+		return err;
+	}
+
+	return opt3001_set_higher_th(dev, 0xBFFF);
```
Comment (MaureenHelm) [inline drivers/sensor/opt3001/opt3001_trigger.c:164]: Where does the default threshold come from? You should probably use a devicetree property to set it

**Comment by cfriedt** [comment]
@GiulianoFranchetto - would have been nice to see this be merged. Do you have any bandwidth to address Maureen's concerns?

**Comment by GiulianoFranchetto** [comment]
@cfriedt Hi. For now, I don't have any time to work on that PR.
For the second comment, the thresholds are defined in the datasheet to disable interrupts. No need to use device-tree for that
